torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,375 @@
+    its current output: the parametrized tensor.
+    If ``leave_parametrized == False``, ``module[tensor_name]`` will be set to
+    its unparametrized value, that is, ``module[tensor_name + "_param"].original()``
+
+    .. warning ::
+
+        If the parametrization changes the size of the tensor and the parametrization
+        is on a parameter being optimized, since this function will register a new
+        parameter, the parameters on the optimizer have to be manually updated via
+        ``optim.params = model.parameters()`` after calling this method.
+
+    Args:
+        module (nn.Module): module from which remove the parametrization
+        tensor_name (str): name of the parametrization to be removed
+        leave_parametrized (bool, optional): leave the attribute ``tensor_name``
+        parametrized or not. Default: False
+    """
+
+    param = getattr(module, tensor_name + "_param")
+    orig = param.original()
+    # Parametrization on a parameter or a buffer
+    is_parameter = isinstance(orig, Parameter)
+    if leave_parametrized:
+        t = getattr(module, tensor_name)
+        if t.size() != orig.size():
+            if is_parameter:
+                orig = Parameter(t)
+            else:
+                orig = t
+        else:
+            orig.data = t
+
+    # Delete the property that manages the parametrization
+    delattr(module.__class__, tensor_name)
+    # Delete the parametrization
+    delattr(module, tensor_name + "_param")
+    if hasattr(module, tensor_name + "_cache"):
+        delattr(module, tensor_name + "_cache")
+
+    if is_parameter:
+        module.register_parameter(tensor_name, orig)
+    else:
+        module.register_buffer(tensor_name, orig)
+
+    # Roll back the fancy parametrized class if no other
+    # buffer or parameter is currently parametrized
+    if not is_parametrized(module):
+        parents = module.__class__.__bases__
+        if len(parents) != 1:
+            raise TypeError("Found a Parametrized module with more than "
+                            "one parent class. This is currently not supported.")
+        module.__class__ = parents[0]
+
+
+class cached:
+    r"""Context-manager that enables the caching system within
+    :class:`torch.nn.Parametrization`
+
+    This is usful when one uses certain parametrized parameter more than
+    once. An example of this is the loop in an RNN model
+    """
+
+    def __init__(self, model):
+        if model is not None and not isinstance(model, Module):
+            raise TypeError("Model should be an nn.Module. Found '{}'"
+                            .format(model.__class__))
+        self.model = model
+
+    def __enter__(self):
+        def _enable_caching(module):
+            for name, param in active_parametrizations(module):
+                if has_caching(module, name):
+                    setattr(module, name + "_cache", param.evaluate())
+        self.model.apply(_enable_caching)
```

Comment:
`.apply` will generally be quite expensive and it's always best to avoid it in the hot path. I'm wondering if there's any reason why would someone want to enable caching on a single module and not on a whole block of code (and then we could use a thread-local property that's cheap to enable). Thoughts?

Reply:
I do not know of any use-case where someone would want to do that. I'd be keen on going with the latter option. Could you elaborate on how to implement it?

Reply:
You could make it a global or a thread-local flag that's stored in this module.

Reply:
I thought a bit more about this one, and I do not think it is possible to do it like this. This `apply` basically signals the parametrizations when to recompute the cache. These have to be done once per module for every module in the network.
This function is called just twice per minibatch, so I do not think it's that problematic. If one wanted to optimise it, we also offer the more local decorator:

```
class ParametrizedRNNCell: pass
class RNN:
  @parametrized_method
  def forward(self, xs):
    out = None
    for x in xs:
      out = self.rnn(x, out)
    return out
```
which may be activated in the module that uses a parametrized object several times.

This could also be done by exposing two methods like `activate_caching` and `invalidate_caching` that could be executed manually for every parametrized layer at the beginning and end of each iteration in the training loop, in order to avoid the `apply` call.

Reply:
> I thought a bit more about this one, and I do not think it is possible to do it like this. 

Can you elaborate on why? I don't see any issues.

Reply:
Consider the example:
```
def Net(nn.Module):
    def __init__(self):
        self.a = nn.Parameter(...)
        self.b = nn.Parameter(...)
        P.register_parametrization(self, "a", MyParam())
        P.register_parametrization(self, "b", MyParam())

    def forward(self, input_):
        # 1.a use a
        # 2.a use a
        # 1.b use b
        # 3.a use a
        # 2.b use b

net = Net()

for batch in batches:
    with cached(net) as net:
        out = net(batch)
```
How would you implement cached so that in two different batches the cache is only updated in 1.a and 1.b, but not in any of the other calls? I truly do not see how to do it...

Reply:
How about this:
```python
_cache_enabled = 0
_cache = {}

@contextmanager
def cache():
  _cache_enabled += 1
  try:
    yield
  finally:
    _cache_enabled -= 1
    if not _cache_enabled:
      _cache = {}


class Parametrization(nn.Module):
  def forward(self, X):
    if not _cache_enabled:
      return self.evaluate(X)
    if id(self) not in _cache:
      _cache[id(self)] = self.evaluate(X)
    return _cache[id(self)]

  def __del__(self):
    _cache.pop(id(self))
```